## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [272]:
import pandas as pd
import numpy as np

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [273]:
# View the column names in the data
print(df.columns)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [274]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [275]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.dtypes

first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object

In [276]:
# What three item categories had the most entries?
df['category'].value_counts().head(3)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [277]:
# For the category with the most entries, which subcategory had the most entries?
category_df = df[df['category'] == 'consumables']

In [278]:
category_df['category'].value_counts().head

<bound method NDFrame.head of category
consumables    23538
Name: count, dtype: int64>

In [279]:
# Which five clients had the most entries in the data?
df = df.sort_values(by="qty", ascending=False)
df = df.reset_index(drop=True)
df.head()

first      last                                  job  \
0    Julie  Anderson                  Exhibition designer   
1      Amy   Stewart            Psychologist, educational   
2  Randall    Watson                    Ceramics designer   
3   Dennis   Hawkins  Social research officer, government   
4   Amanda   Burnett                      Records manager   

                   phone                                email  client_id  \
0             5933099462        anderson_better_4704@bell.biz      78965   
1     979.793.4884x57205               stewartamy@francis.com      86721   
2             3478635175  watson.hospital@scott-mcdonald.info      30143   
3  001-214-888-5838x7410          hawkins.dennis.5195@lee.com      85931   
4     (959)858-6180x6708            evidence.amanda@clark.biz      77354   

   order_id  order_date  order_week  order_year         item_id     category  \
0   3833568  2023-04-29          17        2023  XAE75138-91-2E    furniture   
1   6057452  2023-03-21          12        2023  CCU49971-42-9U    furniture   
2   3566639  2023-02-21           8        2023  XDA18116-89-4A  consumables   
3   7263857  2023-03-05           9        2023  XXM82450-88-9M  consumables   
4   3089707  2023-04-17          16        2023  BAB21028-35-2Z  consumables   

        subcategory  unit_price  unit_cost  unit_weight      qty  line_number  
0            stands      147.66     100.11         0.09  3958244            1  
1            tables      100.40      73.05         9.22   977121            3  
2              pens       24.95      15.09         1.49   842581            2  
3    paper products      323.42     183.84         2.21   637318            3  
4  kitchen supplies       81.23      69.22         6.64   494692            1

In [280]:
# Store the client ids of those top 5 clients in a list.
Top_5_Clients_df = df['client_id'].value_counts()
Top_5_Clients_df.head()


client_id
33615    220
66037    211
46820    209
24741    207
38378    207
Name: count, dtype: int64

In [281]:
top_five_client_ids = Top_5_Clients_df.head(5).index.tolist()

top_five_client_ids

[33615, 66037, 46820, 24741, 38378]

In [282]:
# How many total units (the qty column) did the client with the most entries order order?
highest_client = df['client_id'].value_counts().max()
highest_client

220

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [283]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df ["unit_price"] * df ["qty"]
df.head()

first      last                                  job  \
0    Julie  Anderson                  Exhibition designer   
1      Amy   Stewart            Psychologist, educational   
2  Randall    Watson                    Ceramics designer   
3   Dennis   Hawkins  Social research officer, government   
4   Amanda   Burnett                      Records manager   

                   phone                                email  client_id  \
0             5933099462        anderson_better_4704@bell.biz      78965   
1     979.793.4884x57205               stewartamy@francis.com      86721   
2             3478635175  watson.hospital@scott-mcdonald.info      30143   
3  001-214-888-5838x7410          hawkins.dennis.5195@lee.com      85931   
4     (959)858-6180x6708            evidence.amanda@clark.biz      77354   

   order_id  order_date  order_week  order_year         item_id     category  \
0   3833568  2023-04-29          17        2023  XAE75138-91-2E    furniture   
1   6057452  2023-03-21          12        2023  CCU49971-42-9U    furniture   
2   3566639  2023-02-21           8        2023  XDA18116-89-4A  consumables   
3   7263857  2023-03-05           9        2023  XXM82450-88-9M  consumables   
4   3089707  2023-04-17          16        2023  BAB21028-35-2Z  consumables   

        subcategory  unit_price  unit_cost  unit_weight      qty  line_number  \
0            stands      147.66     100.11         0.09  3958244            1   
1            tables      100.40      73.05         9.22   977121            3   
2              pens       24.95      15.09         1.49   842581            2   
3    paper products      323.42     183.84         2.21   637318            3   
4  kitchen supplies       81.23      69.22         6.64   494692            1   

       subtotal  
0  5.844743e+08  
1  9.810295e+07  
2  2.102240e+07  
3  2.061214e+08  
4  4.018383e+07

In [284]:
import pandas as pd


In [285]:
# # Create a column for shipping price.
# # Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.

df['shipping_price'] = np.where(df['unit_weight'] > 50, df['unit_weight'] * 7, df['unit_weight'] * 10)
df

first      last                                  job  \
0          Julie  Anderson                  Exhibition designer   
1            Amy   Stewart            Psychologist, educational   
2        Randall    Watson                    Ceramics designer   
3         Dennis   Hawkins  Social research officer, government   
4         Amanda   Burnett                      Records manager   
...          ...       ...                                  ...   
54634     Philip     Gates                         Photographer   
54635     Daniel    Bowman                        Archaeologist   
54636      Carla   Vazquez                        Site engineer   
54637  Katherine    Hardin                  Magazine journalist   
54638      Jared    Taylor                        Haematologist   

                        phone                                email  client_id  \
0                  5933099462        anderson_better_4704@bell.biz      78965   
1          979.793.4884x57205               stewartamy@francis.com      86721   
2                  3478635175  watson.hospital@scott-mcdonald.info      30143   
3       001-214-888-5838x7410          hawkins.dennis.5195@lee.com      85931   
4          (959)858-6180x6708            evidence.amanda@clark.biz      77354   
...                       ...                                  ...        ...   
54634        247.214.9743x760             gatesphilip276@foley.com      75947   
54635    001-716-757-9724x862            danielbowman@thompson.com      72892   
54636  001-908-404-5587x64031    vazquez.c.9803@santana-lewis.info      93794   
54637            229-575-6791               hardink5344@lewis.info      78465   
54638            729-249-1032                   taylorj@romero.com      37715   

       order_id  order_date  order_week  order_year         item_id  \
0       3833568  2023-04-29          17        2023  XAE75138-91-2E   
1       6057452  2023-03-21          12        2023  CCU49971-42-9U   
2       3566639  2023-02-21           8        2023  XDA18116-89-4A   
3       7263857  2023-03-05           9        2023  XXM82450-88-9M   
4       3089707  2023-04-17          16        2023  BAB21028-35-2Z   
...         ...         ...         ...         ...             ...   
54634   1286765  2023-02-08           6        2023  UDC75479-26-5A   
54635   6514417  2023-05-17          20        2023  EZC20385-63-6E   
54636   4877398  2023-05-03          18        2023  CXD77584-10-1E   
54637   7520308  2023-02-06           6        2023  DAC03696-47-4B   
54638   4250945  2023-04-16          15        2023  BUZ87260-13-9U   

          category       subcategory  unit_price  unit_cost  unit_weight  \
0        furniture            stands      147.66     100.11         0.09   
1        furniture            tables      100.40      73.05         9.22   
2      consumables              pens       24.95      15.09         1.49   
3      consumables    paper products      323.42     183.84         2.21   
4      consumables  kitchen supplies       81.23      69.22         6.64   
...            ...               ...         ...        ...          ...   
54634     software        accounting      277.01     174.45         4.51   
54635  electronics           cabling       96.86      74.98         0.24   
54636  consumables    paper products      253.71     160.50         4.08   
54637    furniture            stands      169.81     157.24         2.62   
54638    furniture           seating       34.02      20.43        19.89   

           qty  line_number      subtotal  shipping_price  
0      3958244            1  5.844743e+08             0.9  
1       977121            3  9.810295e+07            92.2  
2       842581            2  2.102240e+07            14.9  
3       637318            3  2.061214e+08            22.1  
4       494692            1  4.018383e+07            66.4  
...        ...          ...           ...             ...  
54634        0            2  0.000000e+00      

In [286]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
sales_tax = df['subtotal'] * 0.0925
df['total_price'] = df['subtotal'] + df['shipping_price'] + sales_tax
df

first      last                                  job  \
0          Julie  Anderson                  Exhibition designer   
1            Amy   Stewart            Psychologist, educational   
2        Randall    Watson                    Ceramics designer   
3         Dennis   Hawkins  Social research officer, government   
4         Amanda   Burnett                      Records manager   
...          ...       ...                                  ...   
54634     Philip     Gates                         Photographer   
54635     Daniel    Bowman                        Archaeologist   
54636      Carla   Vazquez                        Site engineer   
54637  Katherine    Hardin                  Magazine journalist   
54638      Jared    Taylor                        Haematologist   

                        phone                                email  client_id  \
0                  5933099462        anderson_better_4704@bell.biz      78965   
1          979.793.4884x57205               stewartamy@francis.com      86721   
2                  3478635175  watson.hospital@scott-mcdonald.info      30143   
3       001-214-888-5838x7410          hawkins.dennis.5195@lee.com      85931   
4          (959)858-6180x6708            evidence.amanda@clark.biz      77354   
...                       ...                                  ...        ...   
54634        247.214.9743x760             gatesphilip276@foley.com      75947   
54635    001-716-757-9724x862            danielbowman@thompson.com      72892   
54636  001-908-404-5587x64031    vazquez.c.9803@santana-lewis.info      93794   
54637            229-575-6791               hardink5344@lewis.info      78465   
54638            729-249-1032                   taylorj@romero.com      37715   

       order_id  order_date  order_week  order_year  ...     category  \
0       3833568  2023-04-29          17        2023  ...    furniture   
1       6057452  2023-03-21          12        2023  ...    furniture   
2       3566639  2023-02-21           8        2023  ...  consumables   
3       7263857  2023-03-05           9        2023  ...  consumables   
4       3089707  2023-04-17          16        2023  ...  consumables   
...         ...         ...         ...         ...  ...          ...   
54634   1286765  2023-02-08           6        2023  ...     software   
54635   6514417  2023-05-17          20        2023  ...  electronics   
54636   4877398  2023-05-03          18        2023  ...  consumables   
54637   7520308  2023-02-06           6        2023  ...    furniture   
54638   4250945  2023-04-16          15        2023  ...    furniture   

            subcategory unit_price  unit_cost  unit_weight      qty  \
0                stands     147.66     100.11         0.09  3958244   
1                tables     100.40      73.05         9.22   977121   
2                  pens      24.95      15.09         1.49   842581   
3        paper products     323.42     183.84         2.21   637318   
4      kitchen supplies      81.23      69.22         6.64   494692   
...                 ...        ...        ...          ...      ...   
54634        accounting     277.01     174.45         4.51        0   
54635           cabling      96.86      74.98         0.24        0   
54636    paper products     253.71     160.50         4.08        0   
54637            stands     169.81     157.24         2.62        0   
54638           seating      34.02      20.43        19.89        0   

       line_number      subtotal  shipping_price   total_price  
0                1  5.844743e+08             0.9  6.385382e+08  
1                3  9.810295e+07            92.2  1.071776e+08  
2                2  2.102240e+07            14.9  2.296698e+07  
3                3  2.061214e+08            22.1  2.251876e+08  
4                1  4.018383e+07            66.4  4.390090e+07  
...            ...           ...             ...           ...  
54634            2  0.000000e+00            45.1

In [287]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']
df

first      last                                  job  \
0          Julie  Anderson                  Exhibition designer   
1            Amy   Stewart            Psychologist, educational   
2        Randall    Watson                    Ceramics designer   
3         Dennis   Hawkins  Social research officer, government   
4         Amanda   Burnett                      Records manager   
...          ...       ...                                  ...   
54634     Philip     Gates                         Photographer   
54635     Daniel    Bowman                        Archaeologist   
54636      Carla   Vazquez                        Site engineer   
54637  Katherine    Hardin                  Magazine journalist   
54638      Jared    Taylor                        Haematologist   

                        phone                                email  client_id  \
0                  5933099462        anderson_better_4704@bell.biz      78965   
1          979.793.4884x57205               stewartamy@francis.com      86721   
2                  3478635175  watson.hospital@scott-mcdonald.info      30143   
3       001-214-888-5838x7410          hawkins.dennis.5195@lee.com      85931   
4          (959)858-6180x6708            evidence.amanda@clark.biz      77354   
...                       ...                                  ...        ...   
54634        247.214.9743x760             gatesphilip276@foley.com      75947   
54635    001-716-757-9724x862            danielbowman@thompson.com      72892   
54636  001-908-404-5587x64031    vazquez.c.9803@santana-lewis.info      93794   
54637            229-575-6791               hardink5344@lewis.info      78465   
54638            729-249-1032                   taylorj@romero.com      37715   

       order_id  order_date  order_week  order_year  ...       subcategory  \
0       3833568  2023-04-29          17        2023  ...            stands   
1       6057452  2023-03-21          12        2023  ...            tables   
2       3566639  2023-02-21           8        2023  ...              pens   
3       7263857  2023-03-05           9        2023  ...    paper products   
4       3089707  2023-04-17          16        2023  ...  kitchen supplies   
...         ...         ...         ...         ...  ...               ...   
54634   1286765  2023-02-08           6        2023  ...        accounting   
54635   6514417  2023-05-17          20        2023  ...           cabling   
54636   4877398  2023-05-03          18        2023  ...    paper products   
54637   7520308  2023-02-06           6        2023  ...            stands   
54638   4250945  2023-04-16          15        2023  ...           seating   

      unit_price unit_cost  unit_weight      qty  line_number      subtotal  \
0         147.66    100.11         0.09  3958244            1  5.844743e+08   
1         100.40     73.05         9.22   977121            3  9.810295e+07   
2          24.95     15.09         1.49   842581            2  2.102240e+07   
3         323.42    183.84         2.21   637318            3  2.061214e+08   
4          81.23     69.22         6.64   494692            1  4.018383e+07   
...          ...       ...          ...      ...          ...           ...   
54634     277.01    174.45         4.51        0            2  0.000000e+00   
54635      96.86     74.98         0.24        0            6  0.000000e+00   
54636     253.71    160.50         4.08        0            4  0.000000e+00   
54637     169.81    157.24         2.62        0            6  0.000000e+00   
54638      34.02     20.43        19.89        0            0  0.000000e+00   

       shipping_price   total_price     line_cost  
0                 0.9  6.385382e+08  3.962598e+08  
1                92.2  1.071776e+08  7.137878e+07  
2                14.9  2.296698e+07  1.271456e+07  
3                22.1  2.251876e+08  1.171646e+08  
4                66.4  4.390090e+07  3.424265e+07  
...               ...           ...

In [288]:
# Create a column for the profit of each line using line cost and line price
df['line_profit'] = df['total_price'] - df['line_cost']
df

first      last                                  job  \
0          Julie  Anderson                  Exhibition designer   
1            Amy   Stewart            Psychologist, educational   
2        Randall    Watson                    Ceramics designer   
3         Dennis   Hawkins  Social research officer, government   
4         Amanda   Burnett                      Records manager   
...          ...       ...                                  ...   
54634     Philip     Gates                         Photographer   
54635     Daniel    Bowman                        Archaeologist   
54636      Carla   Vazquez                        Site engineer   
54637  Katherine    Hardin                  Magazine journalist   
54638      Jared    Taylor                        Haematologist   

                        phone                                email  client_id  \
0                  5933099462        anderson_better_4704@bell.biz      78965   
1          979.793.4884x57205               stewartamy@francis.com      86721   
2                  3478635175  watson.hospital@scott-mcdonald.info      30143   
3       001-214-888-5838x7410          hawkins.dennis.5195@lee.com      85931   
4          (959)858-6180x6708            evidence.amanda@clark.biz      77354   
...                       ...                                  ...        ...   
54634        247.214.9743x760             gatesphilip276@foley.com      75947   
54635    001-716-757-9724x862            danielbowman@thompson.com      72892   
54636  001-908-404-5587x64031    vazquez.c.9803@santana-lewis.info      93794   
54637            229-575-6791               hardink5344@lewis.info      78465   
54638            729-249-1032                   taylorj@romero.com      37715   

       order_id  order_date  order_week  order_year  ... unit_price unit_cost  \
0       3833568  2023-04-29          17        2023  ...     147.66    100.11   
1       6057452  2023-03-21          12        2023  ...     100.40     73.05   
2       3566639  2023-02-21           8        2023  ...      24.95     15.09   
3       7263857  2023-03-05           9        2023  ...     323.42    183.84   
4       3089707  2023-04-17          16        2023  ...      81.23     69.22   
...         ...         ...         ...         ...  ...        ...       ...   
54634   1286765  2023-02-08           6        2023  ...     277.01    174.45   
54635   6514417  2023-05-17          20        2023  ...      96.86     74.98   
54636   4877398  2023-05-03          18        2023  ...     253.71    160.50   
54637   7520308  2023-02-06           6        2023  ...     169.81    157.24   
54638   4250945  2023-04-16          15        2023  ...      34.02     20.43   

      unit_weight      qty  line_number      subtotal  shipping_price  \
0            0.09  3958244            1  5.844743e+08             0.9   
1            9.22   977121            3  9.810295e+07            92.2   
2            1.49   842581            2  2.102240e+07            14.9   
3            2.21   637318            3  2.061214e+08            22.1   
4            6.64   494692            1  4.018383e+07            66.4   
...           ...      ...          ...           ...             ...   
54634        4.51        0            2  0.000000e+00            45.1   
54635        0.24        0            6  0.000000e+00             2.4   
54636        4.08        0            4  0.000000e+00            40.8   
54637        2.62        0            6  0.000000e+00            26.2   
54638       19.89        0            0  0.000000e+00           198.9   

        total_price     line_cost   line_profit  
0      6.385382e+08  3.962598e+08  2.422784e+08  
1      1.071776e+08  7.137878e+07  3.579878e+07  
2      2.296698e+07  1.271456e+07  1.025242e+07  
3      2.251876e+08  1.171646e+08  1.080231e+08  
4      4.390090e+07  3.424265e+07  9.658255e+06  
...             ...           ...           ...  
54634  4.510000e+01  4.510000e+01

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [289]:
# # Check your work using the totals above
order_totals = df.groupby('order_id')['total_price'].sum().reset_index()
receipt_totals = {
    2742071: 152811.89,
    2173913: 162388.71,
    6128929: 923441.25
}
for order_id, receipt_total in receipt_totals.items():
    calculated_total = order_totals.loc[order_totals['order_id'] == order_id, 'total_price'].values[0]
    if calculated_total == receipt_total:
        print(f"Order ID {order_id}: Calculated total matches the receipt total of ${receipt_total:.2f}")
    else:
        print(f"Order ID {order_id}: Calculated total of ${calculated_total:.2f} does NOT match the receipt total of ${receipt_total:.2f}")

Order ID 2742071: Calculated total of $139825.57 does NOT match the receipt total of $152811.89
Order ID 2173913: Calculated total of $119815.13 does NOT match the receipt total of $162388.71
Order ID 6128929: Calculated total of $752707.06 does NOT match the receipt total of $923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [290]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
top_clients_by_qty = df.groupby('client_id')['qty'].sum().nlargest(5)
total_spend_by_top_clients = df[df['client_id'].isin(top_clients_by_qty.index)].groupby('client_id')['total_price'].sum()
total_spend_by_top_clients

client_id
30143    2.351828e+07
77354    4.592767e+07
78965    6.500963e+08
85931    2.263609e+08
86721    1.109857e+08
Name: total_price, dtype: float64

In [291]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
top_clients = df.groupby('client_id')['qty'].sum().nlargest(5).index
summary_data = df[df['client_id'].isin(top_clients)].groupby('client_id').agg(
    total_units_purchased=pd.NamedAgg(column='qty', aggfunc='sum'),
    total_shipping_price=pd.NamedAgg(column='shipping_price', aggfunc='sum'),
    total_revenue=pd.NamedAgg(column='total_price', aggfunc='sum'),
    total_profit=pd.NamedAgg(column='line_profit', aggfunc='sum')
)
summary_data

total_units_purchased  total_shipping_price  total_revenue  \
client_id                                                               
30143                     846069                2157.3   2.351828e+07   
77354                     506382                2564.3   4.592767e+07   
78965                    4033302                7383.3   6.500963e+08   
85931                     644423                2763.0   2.263609e+08   
86721                    1010426                6154.6   1.109857e+08   

           total_profit  
client_id                
30143      1.041956e+07  
77354      1.022940e+07  
78965      2.468588e+08  
85931      1.083578e+08  
86721      3.702364e+07

In [292]:
# Format the data and rename the columns to names suitable for presentation.

# Define the Money Columns
money_columns = ['total_revenue', 'total_profit']

# Define a Function That Converts a Dollar Amount to Millions
def currency_format_millions(amount):
    return f"${amount / 1_000_000:.2f}M"

# Apply the Conversion Function to Only the Money Columns
for column in money_columns:
    summary_data[column] = summary_data[column].apply(currency_format_millions)

# Rename the Columns to Reflect the Change in the Money Format
summary_data.rename(columns={
    'total_units_purchased': 'Total Units',
    'total_shipping_price': 'Total Shipping (M$)',
    'total_revenue': 'Total Revenue (M$)',
    'total_profit': 'Total Profit (M$)'
}, inplace=True)

summary_data


Total Units  Total Shipping (M$) Total Revenue (M$)  \
client_id                                                        
30143           846069               2157.3            $23.52M   
77354           506382               2564.3            $45.93M   
78965          4033302               7383.3           $650.10M   
85931           644423               2763.0           $226.36M   
86721          1010426               6154.6           $110.99M   

          Total Profit (M$)  
client_id                    
30143               $10.42M  
77354               $10.23M  
78965              $246.86M  
85931              $108.36M  
86721               $37.02M

In [293]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_data = summary_data.sort_values('Total Profit (M$)', ascending=False)

sorted_data

Total Units  Total Shipping (M$) Total Revenue (M$)  \
client_id                                                        
86721          1010426               6154.6           $110.99M   
78965          4033302               7383.3           $650.10M   
85931           644423               2763.0           $226.36M   
30143           846069               2157.3            $23.52M   
77354           506382               2564.3            $45.93M   

          Total Profit (M$)  
client_id                    
86721               $37.02M  
78965              $246.86M  
85931              $108.36M  
30143               $10.42M  
77354               $10.23M